File: mobile/Chela/LoginGateConnect.swift
import Foundation

/// Connect pressed on the Control UI's OWN login gate, read from
/// core/spec/login-gate-connect.json.
///
/// The desktop reads the same file through core/login-gate-connect.js. This
/// client has no Swift copy of the script: it bundles the spec and installs those
/// bytes, so the two clients watch for the press the same way. The reports arrive
/// on the pairing observer's channel (PairingBridge), and PageCover answers
/// them: the loading screen at once on the press, lifted once the interface has
/// rendered, and its failed state on a refusal or at the deadline.
enum LoginGateConnect {
    private struct Spec: Decodable {
        struct Reports: Decodable {
            let pressed: String
            let rendered: String
            let failed: String
        }
        let reports: Reports
        let deadlineMs: Int
        let hook: [String]
    }

    private static let spec: Spec? = try? BundledSpec.load("login-gate-connect", as: Spec.self)

    /// The injected script. Nil when the build did not bundle the spec, which
    /// LoginGateConnectParityTests fails on.
    static var script: String? {
        guard let lines = spec?.hook, !lines.isEmpty else { return nil }
        return lines.joined(separator: "\n")
    }

    /// How long a press may go unanswered before the cover lands on its failed
    /// state.
    static var deadlineMs: Int { spec?.deadlineMs ?? 15000 }

    /// The three kinds the script posts.
    enum Report: Equatable {
        case pressed
        case rendered
        case failed(title: String)
    }

    /// Read a posted body into a report, or nil. The payload is the page's, so
    /// only the three kinds the spec names are accepted and the title is narrowed.
    static func read(_ body: [String: Any]) -> Report? {
        guard let reports = spec?.reports, let kind = body["kind"] as? String else { return nil }
        switch kind {
        case reports.pressed: return .pressed
        case reports.rendered: return .rendered
        case reports.failed:
            let raw = (body["title"] as? String) ?? ""
            let squeezed = raw.split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
            return .failed(title: String(squeezed.prefix(160)))
        default: return nil
        }
    }

    static let decodedKeys: Set<String> = ["reports", "deadlineMs", "hook"]
    static let ignoredKeys: Set<String> = ["description", "why", "selectors", "upstreamComponent"]
}

File: core/login-gate-connect.js
// Connect pressed on the Control UI's own login gate, and the loading cover that
// answers both that press and the gate's own presence.
//
// The gate is upstream's page and is not patched: a script installed at document
// start (spec.hook) reports the press, the gate leaving, the gate's failure coming
// back and -- because the reader should never see that screen at all -- the gate's
// arrival and departure. This module turns those reports into the cover. See
// core/spec/login-gate-connect.json for why each part is shaped as it is, and
// core/ui/CONVENTIONS.md for the rules it serves: the loading screen goes up rather
// than the gate, comes down only once the interface has rendered, and a connect
// that does not answer lands on the failed state with Try again on a bounded
// deadline.
//
// Every client uses this: the desktop installs LOGIN_GATE_CONNECT_SCRIPT beside
// the pairing observer, and the iOS client bundles the spec and runs the same
// bytes (mobile/Chela/LoginGateConnect.swift).

import spec from './spec/login-gate-connect.json' with { type: 'json' };

/** The injected script, joined from the spec's lines. */
export const LOGIN_GATE_CONNECT_SCRIPT = spec.hook.join('\n');

/** The report kinds the script posts. */
export const GATE_REPORTS = Object.freeze({ ...spec.reports });

/** How long a press may go unanswered before the cover lands on its failed state. */
export const CONNECT_DEADLINE_MS = spec.deadlineMs;

/** How long the cover is held over a page that is on its own gate before it fails. */
export const GATE_HOLD_MS = spec.gateMs;

/**
 * Read a report from the page into a gate event, or null.
 *
 * The payload crosses a page boundary, so only the kinds the spec names are
 * accepted, and the one free-text field is narrowed to a short string.
 *
 * @param {unknown} body a parsed report
 * @returns {{event: 'pressed'|'rendered'|'failed'|'gateShown'|'gateGone'|'pageReady', title: string}|null}
 */
export function readGateReport(body) {
  if (!body || typeof body !== 'object') return null;
  const event = Object.keys(GATE_REPORTS).find((key) => GATE_REPORTS[key] === body.kind);
  if (!event) return null;
  const title = typeof body.title === 'string' ? body.title.replace(/\s+/g, ' ').trim().slice(0, 160) : '';
  return { event, title };
}

/**
 * The host's half: one cover, and the endings that may take it down.
 *
 * Two things raise it: the press on the gate's own Connect, and the gate itself
 * appearing. One thing lifts it: an answer -- the interface rendered, a failure the
 * page reports, or, for a gate the reader was never shown, that gate going away.
 * Both are bounded, the press at deadlineMs and the gate at gateMs, so neither can
 * leave a cover that never lets go (core/ui/CONVENTIONS.md).
 *
 * A press while the gate is up is the gate's own Connect, and takes the cover over
 * from it: a press while one is already in flight is the same attempt, and cancel()
 * drops the press for a load of the client's own that supersedes it.
 *
 * `pageReady` is the RESET rather than an ending: a gate reported by one document
 * says nothing about the next, so a document beginning is what forgets it. Without
 * that, a gate that was up when a new page began would keep the cover down over a
 * page that is connecting normally.
 *
 * Clock-free, so every timing is exercised from one test run.
 *
 * @param {{cover: () => void, lift: (why: string) => void, fail: (why: string, title: string) => void,
 *   deadlineMs?: number, gateMs?: number, setTimer?: Function, clearTimer?: Function}} deps
 */
export function createLoginGateCover({ cover, lift, fail, deadlineMs = CONNECT_DEADLINE_MS, gateMs = GATE_HOLD_MS, setTimer = setTimeout, clearTimer = clearTimeout }) {
  let pressTimer = null;
  let gateTimer = null;
  let active = false;
  let onGate = false;
  const forgetGate = () => {
    if (gateTimer !== null) clearTimer(gateTimer);
    gateTimer = null;
    onGate = false;
  };
  const endPress = () => {
    if (pressTimer !== null) clearTimer(pressTimer);
    pressTimer = null;
    const was = active;
    active = false;
    return was;
  };
  return {
    report(report) {
      if (!report) return;
      if (report.event === 'pageReady') { forgetGate(); return; }
      if (report.event === 'gateShown') {
        if (onGate) return;
        onGate = true;
        if (active) return; // the press owns the cover, and has its own deadline
        cover();
        gateTimer = setTimer(() => {
          gateTimer = null;
          if (!onGate || active) return;
          onGate = false;
          fail('gate', report.title || '');
        }, gateMs);
        return;
      }
      if (report.event === 'gateGone') {
        if (!onGate) return;
        const wasPressed = active;
        forgetGate();
        if (wasPressed) return; // the press's own answer ends it
        lift('gate-gone');
        return;
      }
      if (report.event === 'pressed') {
        if (active) return;
        forgetGate();
        active = true;
        cover();
        pressTimer = setTimer(() => { pressTimer = null; if (endPress()) fail('deadline', ''); }, deadlineMs);
        return;
      }
      if (!active) return;
      if (report.event === 'rendered') { forgetGate(); endPress(); lift('rendered'); return; }
      if (report.event === 'failed') { forgetGate(); endPress(); fail('refused', report.title || ''); }
    },
    cancel() { endPress(); forgetGate(); },
    get active() { return active; },
    get onGate() { return onGate; },
  };
}

